[Bug] Snapshotter unit tests which require root permissions are skipped in CI

### Description

7 snapshotter unit tests require root permissions; however, the current CI setup is not running unit tests as root. This means these tests are being skipped. Moreover when running the tests locally with root permissions, the tests are still skipped.

### Steps to reproduce the bug

Run unit tests with verbose output to see the skipped tests:
```
GO_TEST_FLAGS="-v -count=1" make test | grep "SKIP:"
```

Output:
```
--- SKIP: TestRemotePrepare (0.00s)
--- SKIP: TestRemoteView (0.00s)
--- SKIP: TestRemoteOverlay (0.00s)
--- SKIP: TestRemoteCommit (0.00s)
--- SKIP: TestFailureDetection (0.00s)
--- SKIP: TestOverlay (0.00s)
--- SKIP: TestOverlayOverlayRead (0.00s)
```

Run unit tests with root permissions results in the same behavior:
```
GO_TEST_FLAGS="-v -count=1" sudo make test | grep "SKIP:"
```

Output:
```
--- SKIP: TestRemotePrepare (0.00s)
--- SKIP: TestRemoteView (0.00s)
--- SKIP: TestRemoteOverlay (0.00s)
--- SKIP: TestRemoteCommit (0.00s)
--- SKIP: TestFailureDetection (0.00s)
--- SKIP: TestOverlay (0.00s)
--- SKIP: TestOverlayOverlayRead (0.00s)
```

### Describe the results you expected

1. All of the snapshotter's unit tests should be ran in CI.
2. Running unit tests locally with root permissions should run all unit tests.

### Host information

1. OS: Amazon Linux
2. Snapshotter Version: 
```
soci version fe3fc93.m fe3fc9339a3fa5a3f602cfcbc21842bb8d3b9028.m
```
3. Containerd Version: 
```
containerd github.com/containerd/containerd v1.6.24-36-gbf6f562cc bf6f562ccf2734f6f5e5723d043c45bb7d9620ac
```


### Any additional context or information about the bug

The unit tests are taking dependency on `testutil.RequiresRoot` from `github.com/containerd/containerd/pkg/testutil`.

Tested in GitHub Codespaces. At least TestRemotePrepare still needs root.

```
@kzys ➜ /workspaces/soci-snapshotter/snapshot (fix-376) $ uname -a
Linux codespaces-13259f 6.8.0-1044-azure #50~22.04.1-Ubuntu SMP Wed Dec  3 15:13:22 UTC 2025 x86_64 x86_64 x86_64 GNU/Linux
@kzys ➜ /workspaces/soci-snapshotter/snapshot (fix-376) $ go test -v ./... -run TestRemotePrepare
=== RUN   TestRemotePrepare
time="2026-02-25T07:39:32Z" level=info msg="preparing filesystem mount at mountpoint=/tmp/TestRemotePrepare2327511264/001/snapshots/1/fs" key=/tmp/prepareTarget parent=
    snapshot_test.go:400: failed to bind mount "/tmp/TestRemotePrepare2327511264/002" to "/tmp/TestRemotePrepare2327511264/001/snapshots/1/fs": operation not permitted
--- FAIL: TestRemotePrepare (0.00s)
FAIL
FAIL    github.com/awslabs/soci-snapshotter/snapshot    0.009s
FAIL
```

Regarding testutil.RequiresRoot, it is from containerd and we need to pass -test.root.

```
@kzys ➜ /workspaces/soci-snapshotter/snapshot (fix-376) $ go test -exec sudo -test.root -v -run TestRemotePrepare 
=== RUN   TestRemotePrepare
INFO[0000] preparing filesystem mount at mountpoint=/tmp/TestRemotePrepare1278958485/001/snapshots/1/fs  key=/tmp/prepareTarget parent=
INFO[0000] remote snapshot successfully prepared.        key=/tmp/prepareTarget parent= remote-snapshot-prepared=true
--- PASS: TestRemotePrepare (0.01s)
PASS
ok      github.com/awslabs/soci-snapshotter/snapshot    0.018s
```

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -47,6 +47,23 @@ jobs:
       - name: Show test coverage
         run: make show-test-coverage
 
+  # This is a temporary workaround to ensure we are testing our packages
+  # that require root (currently only the snapshot package).
+  # ref: https://github.com/awslabs/soci-snapshotter/issues/1878
+  test-root:
+    runs-on: ubuntu-24.04
+    strategy:
+      matrix:
+        go-version: ['1.25.7', '1.26.0']
+    timeout-minutes: 15
+    steps:
+      - uses: actions/checkout@v6
+      - uses: actions/setup-go@v6
+        with:
+          go-version: ${{ matrix.go-version }}
+      # Only snapshot package has tests that require root
+      - run: sudo -E env PATH=$PATH go test ./snapshot -test.root
+
   integration:
     needs: setup
     runs-on: ${{ fromJSON(needs.setup.outputs.runner-labels)[matrix.os] }}

--- a/snapshot/snapshot_test.go
+++ b/snapshot/snapshot_test.go
@@ -43,6 +43,7 @@ import (
 
 	"github.com/awslabs/soci-snapshotter/idtools"
 	"github.com/containerd/containerd/mount"
+	"github.com/containerd/containerd/namespaces"
 	"github.com/containerd/containerd/pkg/testutil"
 	"github.com/containerd/containerd/snapshots"
 	"github.com/containerd/containerd/snapshots/overlay/overlayutils"
@@ -57,9 +58,7 @@ const (
 	brokenLabel              = "containerd.io/snapshot/broken"
 )
 
-func prepareWithTarget(t *testing.T, sn snapshots.Snapshotter, target, key, parent string, labels map[string]string) string {
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
+func prepareWithTarget(ctx context.Context, t *testing.T, sn snapshots.Snapshotter, target, key, parent string, labels map[string]string) string {
 	if labels == nil {
 		labels = make(map[string]string)
 	}
@@ -70,10 +69,18 @@ func prepareWithTarget(t *testing.T, sn snapshots.Snapshotter, target, key, pare
 	return target
 }
 
+func failOnError(t *testing.T, fn func() error) {
+	err := fn()
+	if err != nil {
+		t.Fatalf("error during test cleanup: %v", err)
+	}
+}
+
 func TestRemotePrepare(t *testing.T) {
 	testutil.RequiresRoot(t)
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
+	ctx = namespaces.WithNamespace(ctx, namespaces.Default)
 	root := t.TempDir()
 	sn, err := NewSnapshotter(ctx, root, bindFileSystem(t))
 	if err != nil {
@@ -81,8 +88,8 @@ func TestRemotePrepare(t *testing.T) {
 	}
 
 	// Prepare a remote snapshot.
-	target := prepareWithTarget(t, sn, "testTarget", "/tmp/prepareTarget", "", nil)
-	defer sn.Remove(ctx, target)
+	target := prepareWithTarget(ctx, t, sn, "testTarget", "/tmp/prepareTarget", "", nil)
+	defer failOnError(t, func() error { return sn.Remove(ctx, target) })
 
 	// Get internally committed remote snapshot.
 	var tinfo *snapshots.Info
@@ -122,6 +129,7 @@ func TestRemoteOverlay(t *testing.T) {
 	testutil.RequiresRoot(t)
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
+	ctx = namespaces.WithNamespace(ctx, namespaces.Default)
 	root := t.TempDir()
 	sn, err := NewSnapshotter(ctx, root, bindFileSystem(t))
 	if err != nil {
@@ -129,8 +137,8 @@ func TestRemoteOverlay(t *testing.T) {
 	}
 
 	// Prepare a remote snapshot.
-	target := prepareWithTarget(t, sn, "testTarget", "/tmp/prepareTarget", "", nil)
-	defer sn.Remove(ctx, target)
+	target := prepareWithTarget(ctx, t, sn, "testTarget", "/tmp/prepareTarget", "", nil)
+	defer failOnError(t, func() error { return sn.Remove(ctx, target) })
 
 	// Prepare a new layer based on the remote snapshot.
 	pKey := "/tmp/test"
@@ -138,6 +146,7 @@ func TestRemoteOverlay(t *testing.T) {
 	if err != nil {
 		t.Fatalf("faild to prepare using lower remote layer: %v", err)
 	}
+	defer failOnError(t, func() error { return sn.Remove(ctx, pKey) })
 	if len(mounts) != 1 {
 		t.Errorf("should only have 1 mount but received %d", len(mounts))
 	}
@@ -178,6 +187,7 @@ func TestRemoteCommit(t *testing.T) {
 	testutil.RequiresRoot(t)
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
+	ctx = namespaces.WithNamespace(ctx, namespaces.Default)
 	root := t.TempDir()
 	sn, err := NewSnapshotter(ctx, root, bindFileSystem(t))
 	if err != nil {
@@ -185,8 +195,8 @@ func TestRemoteCommit(t *testing.T) {
 	}
 
 	// Prepare a remote snapshot.
-	target := prepareWithTarget(t, sn, "testTarget", "/tmp/prepareTarget", "", nil)
-	defer sn.Remove(ctx, target)
+	target := prepareWithTarget(ctx, t, sn, "testTarget", "/tmp/prepareTarget", "", nil)
+	defer failOnError(t, func() error { return sn.Remove(ctx, target) })
 
 	// Prepare a new snapshot based on the remote snapshot
 	pKey := "/tmp/test"
@@ -201,29 +211,34 @@ func TestRemoteCommit(t *testing.T) {
 	if err := m.Mount(snapshot); err != nil {
 		t.Fatal(err)
 	}
-	defer mount.Unmount(snapshot, 0)
+	defer failOnError(t, func() error { return mount.Unmount(snapshot, 0) })
 	if err := os.WriteFile(filepath.Join(snapshot, "bar"), []byte("hi"), 0660); err != nil {
 		t.Fatal(err)
 	}
-	mount.Unmount(snapshot, 0)
+	if err := mount.Unmount(snapshot, 0); err != nil {
+		t.Fatalf("error unmounting: %v", err)
+	}
 
 	// Commit the active snapshot
 	cKey := "/tmp/layer"
 	if err := sn.Commit(ctx, cKey, pKey); err != nil {
 		t.Fatal(err)
 	}
+	defer failOnError(t, func() error { return sn.Remove(ctx, cKey) })
 
 	// Validate the committed snapshot
 	check := t.TempDir()
-	mounts, err = sn.Prepare(ctx, "/tmp/test2", cKey)
+	tmpKey := "/tmp/test2"
+	mounts, err = sn.Prepare(ctx, tmpKey, cKey)
 	if err != nil {
 		t.Fatal(err)
 	}
+	defer failOnError(t, func() error { return sn.Remove(ctx, tmpKey) })
 	m = mounts[0]
 	if err := m.Mount(check); err != nil {
 		t.Fatal(err)
 	}
-	defer mount.Unmount(check, 0)
+	defer failOnError(t, func() error { return mount.Unmount(check, 0) })
 	data, err := os.ReadFile(filepath.Join(check, "bar"))
 	if err != nil {
 		t.Fatal(err)
@@ -309,6 +324,7 @@ func TestFailureDetection(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			ctx, cancel := context.WithCancel(context.Background())
 			defer cancel()
+			ctx = namespaces.WithNamespace(ctx, namespaces.Default)
 			root := t.TempDir()
 			fi := bindFileSystem(t)
 			sn, err := NewSnapshotter(ctx, root, fi)
@@ -332,8 +348,8 @@ func TestFailureDetection(t *testing.T) {
 				if broken {
 					labels[brokenLabel] = "true"
 				}
-				target := prepareWithTarget(t, sn, targetName, key, pKey, labels)
-				defer sn.Remove(ctx, target)
+				target := prepareWithTarget(ctx, t, sn, targetName, key, pKey, labels)
+				defer failOnError(t, func() error { return sn.Remove(ctx, target) })
 				pKey = target
 			}
 
@@ -347,23 +363,23 @@ func TestFailureDetection(t *testing.T) {
 				if err := sn.Commit(ctx, cKey, key); err != nil {
 					t.Fatal(err)
 				}
-				defer sn.Remove(ctx, cKey)
+				defer failOnError(t, func() error { return sn.Remove(ctx, cKey) })
 				pKey = cKey
 			}
 
 			// Tests if we can detect layer unavailablity
 			key := "/tmp/snapshot.test"
 			fs.checkFailure = true
-			defer sn.Remove(ctx, key)
 			if _, err := sn.Prepare(ctx, key, pKey); !check(t, tt.wantOK, err) {
 				return
 			}
+			defer failOnError(t, func() error { return sn.Remove(ctx, key) })
 			fs.checkFailure = false
 			key2 := "/tmp/test2"
 			if _, err = sn.Prepare(ctx, key2, pKey); err != nil {
 				t.Fatal(err)
 			}
-			defer sn.Remove(ctx, key2)
+			defer failOnError(t, func() error { return sn.Remove(ctx, key2) })
 			fs.checkFailure = true
 			if _, err := sn.Mounts(ctx, key2); !check(t, tt.wantOK, err) {
 				return
@@ -496,6 +512,7 @@ func TestOverlay(t *testing.T) {
 func TestOverlayMounts(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
+	ctx = namespaces.WithNamespace(ctx, namespaces.Default)
 	root := t.TempDir()
 	o, _, err := newSnapshotter(ctx, root)
 	if err != nil {
@@ -527,6 +544,7 @@ func TestOverlayMounts(t *testing.T) {
 func TestOverlayCommit(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
+	ctx = namespaces.WithNamespace(ctx, namespaces.Default)
 	root := t.TempDir()
 	o, _, err := newSnapshotter(ctx, root)
 	if err != nil {
@@ -549,6 +567,7 @@ func TestOverlayCommit(t *testing.T) {
 func TestOverlayOverlayMount(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
+	ctx = namespaces.WithNamespace(ctx, namespaces.Default)
 	root := t.TempDir()
 	o, _, err := newSnapshotter(ctx, root)
 	if err != nil {
@@ -630,6 +649,7 @@ func TestOverlayOverlayRead(t *testing.T) {
 	testutil.RequiresRoot(t)
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
+	ctx = namespaces.WithNamespace(ctx, namespaces.Default)
 	root := t.TempDir()
 	o, _, err := newSnapshotter(ctx, root)
 	if err != nil {
@@ -657,7 +677,7 @@ func TestOverlayOverlayRead(t *testing.T) {
 	if err := mount.All(mounts, dest); err != nil {
 		t.Fatal(err)
 	}
-	defer syscall.Unmount(dest, 0)
+	defer failOnError(t, func() error { return syscall.Unmount(dest, 0) })
 	data, err := os.ReadFile(filepath.Join(dest, "foo"))
 	if err != nil {
 		t.Fatal(err)
@@ -670,6 +690,7 @@ func TestOverlayOverlayRead(t *testing.T) {
 func TestOverlayView(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
+	ctx = namespaces.WithNamespace(ctx, namespaces.Default)
 	root := t.TempDir()
 	o, _, err := newSnapshotter(ctx, root)
 	if err != nil {

